test/rules/naming/non-state-vars-leading-underscore.js

```diff
@@ -0,0 +1,76 @@
+const assert = require('assert')
+const { processStr } = require('../../../lib/index')
+const { contractWith } = require('../../common/contract-builder')
+
+const config = {
+  rules: { 'non-state-vars-leading-underscore': 'error' },
+}
+
+describe('non-state-vars-leading-underscore', () => {
+  describe('non state vars should have a leading underscore', function () {
+    ;[
+      {
+        description: 'block variable',
+        correct: contractWith('function foo() public { uint myVar; }'),
+        incorrect: contractWith('function foo() public { uint _myVar; }'),
+      },
+      {
+        description: 'return variable',
+        correct: contractWith('function foo() public returns (uint256 _foo){}'),
+        incorrect: contractWith('function foo() public returns (uint256 foo){}'),
+      },
+      {
+        description: 'function param',
+        correct: contractWith('function foo( uint256 _foo ) public {}'),
+        incorrect: contractWith('function foo( uint256 foo ) public {}'),
+      },
+      {
+        description: 'file-level constant',
+        correct: contractWith('function foo() public returns (uint256 _foo){}'),
+        incorrect: contractWith('function foo() public returns (uint256 foo){}'),
+      },
+    ].forEach(({ description, correct, incorrect }) => {
+      it(`should raise an error if a ${description} does not start with an underscore`, () => {
+        const report = processStr(correct, config)
+        assert.equal(report.errorCount, 1)
+        assert(report.messages[0].message.includes('should start with _'))
+      })
+
+      it(`should NOT raise an error if a ${description} starts with an underscore`, () => {
+        const report = processStr(incorrect, config)
+        assert.equal(report.errorCount, 0)
+      })
+    })
+  })
+
+  describe('state vars should NOT have a leading underscore', function () {
```

Comment:
@0xGorilla is part of the desired semantics of this rule that it reports when a state variable starts with an underscore?

Reply:
@juanpcapurro I don't think so because it would go against the `private-vars-leading-underscore` rule. Some state vars can have underscores in their names, some can't, depending on the visibility of the variable.
For reference: https://github.com/protofire/solhint/blob/e9342aaed71b1ac7c368b4ade6283ceddd192922/lib/rules/naming/private-vars-leading-underscore.js#L140-L155